### PR TITLE
Split apart bin and use config files for constants

### LINKED ISSUE #1834: Move grading .py scripts into proper python module
This will make it easier for us to "properly" link to them within other python scripts and have intellisense work properly within IntelliJ or other such IDEs, as well as clean up `bin` of files that are no longer supposed to be run directly.

The proposal would be move:
* grade_item.py
* grade_items_logging.py
* insert_database_version_data.py
* write_grade_history.py

to `bin/grading` which would also contain a blank `__init__.py` file and then update the import references to include the module name. (so you'd do for example `from grading import write_grade_history`)

### PR DESCRIPTION
**Pull request by MasterOdin**
This PR does a couple things:
- Puts autograder python scripts (that are only included) into a python module folder under `bin/` (named autograder), closes #1834.
- Removes the shebang line and any reference to `if __name__ == '__main__'` line from files under bin/autograder.
- Split apart the root:root 500 files in `bin/` to their own directory `sbin/`.
- Have all scripts under `bin/` and `sbin/` reference the .json files in `config/` instead of relying on replacing strings within the scripts when you run install.
- Separate out the install part of files in `bin/` and `sbin/` into their own sub-helper file (`INSTALL_SUBMITTY_HELPER_BIN.sh`) that is sourced from `INSTALL_SUBMITTY_HELPER.sh`, but can be run on its own.
- Separate out the install part of the files in `site/` into its own sub-helper file (`INSTALL_SUBMITTY_HELPER_SITE.sh`) that is sourced from `INSTALL_SUBMITTY_HELPER.sh`, but can be run on its own.

### CODE REVIEW

**Comment by MasterOdin** [comment]
To get this to work, I changed the ownership of `/usr/local/submitty/config/database.json` to be `root:hwcronphp` and added `www-data` to `hwcronphp` group so that everyone has access to the file that needs it via the group (`hwphp`, `www-data`, and `hwcron`).

**Comment by MasterOdin** [comment]
Changed the permissions for `/usr/local/submitty/config/database.json` to be `www-data:hwcronphp` with `hwcronphp` containing just `hwphp` and `hwcron`.

**Comment by MasterOdin** [comment]
@bmcutler looking at the issue list, we did have some discussion at https://github.com/Submitty/Submitty/issues/1144#issuecomment-318546753 which I think uses a slightly different naming scheme then the one I chose here. While I agree we can do more separation of things (like  #1933), I'd prefer that we do that after this is merged as this blocks a couple of things I'd like to add for ease of development for GSoC and RCOS summer people.

File: .setup/CONFIGURE_SUBMITTY.py [hunk 1]
```diff
@@ -389,7 +389,7 @@ def write(x=''):
 
     with open(DATABASE_JSON, 'w') as json_file:
         json.dump(config, json_file, indent=2)
-    shutil.chown(DATABASE_JSON, HWPHP_USER, 'www-data')
+    shutil.chown(DATABASE_JSON, 'www-data', HWCRONPHP_GROUP)
```
Comment (bmcutler) [inline .setup/CONFIGURE_SUBMITTY.py]: we're getting rid of www-data in a later issue/PR, right?

Comment (MasterOdin) [reply .setup/CONFIGURE_SUBMITTY.py]: Yes, in #1904. Going to leave it for now till that is done (so as to not break git)

**Comment by MasterOdin** [comment]
I moved killall.py, autograder* and authentication.py from bin to sbin. I didn't move generate_repos.py as it's a root only script so it'd be weird to be in bin and following the work done in #1904 and my work in [hwcron_jobs](https://github.com/Submitty/Submitty/tree/hwcron_jobs) branch to develop a daemon for handling jobs that it has permissions to do (like building homeworks), I'd like to probably make the generate_repos.py step be done by that service when an instructor creates a gradeable.

**Comment by bmcutler** [comment]
If you don't move generate_repos.py back to bin, it will likely be broken for instructors right now, since the limited sudo is presumably tied to a specific full path (and also, we'd need to change the docs).

I agree, long term, that should be automated.  Note:  It cannot be completed at creation of gradeable time however.  For team assignments we have to run generate repos after the teams have been formed (it can be run multiple times to handle teams formed since the last time it was run.  And also it would be necessary to re-run to handle newly registered students for individual assignments.

**Comment by bmcutler** [comment]
Evan did a bunch of testing, so I'm going to cross my fingers and hope everything else is good.

File: .setup/INSTALL_SUBMITTY_HELPER_BIN.sh [hunk 2]
```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env bash
+
+########################################################################################################################
+########################################################################################################################
+# COPY VARIOUS SCRIPTS USED BY INSTRUCTORS AND SYS ADMINS FOR COURSE ADMINISTRATION
+
+echo -e "Copy the non-root scripts"
+
+if [ -z ${SUBMITTY_INSTALL_DIR+x} ]; then
```
[the remark this answers is not in the document]
Comment (bmcutler) [reply .setup/INSTALL_SUBMITTY_HELPER_BIN.sh]: why did this get closed as outdated?

File: .setup/INSTALL_SUBMITTY_HELPER_BIN.sh [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+fi
+
+# make the directory (has a different name)
+mkdir -p ${SUBMITTY_INSTALL_DIR}/bin
+chown root:${COURSE_BUILDERS_GROUP} ${SUBMITTY_INSTALL_DIR}/bin
+chmod 755 ${SUBMITTY_INSTALL_DIR}/bin
+
+mkdir -p ${SUBMITTY_INSTALL_DIR}/bin/autograder
+chown root:${HWCRON_USER} ${SUBMITTY_INSTALL_DIR}/bin/autograder
+chmod 755 ${SUBMITTY_INSTALL_DIR}/bin/autograder
```
[the remark this answers is not in the document]
Comment (bmcutler) [reply .setup/INSTALL_SUBMITTY_HELPER_BIN.sh]: whey did this get closed as outdated?

[the remark this answers is not in the document]
Comment (MasterOdin) [reply .setup/INSTALL_SUBMITTY_HELPER_BIN.sh]: It's probably because you're commenting on a diff of the non-latest commit (which I had changed things slightly).

Did this though.

File: .setup/INSTALL_SUBMITTY_HELPER_BIN.sh [hunk 4]
```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env bash
+
+########################################################################################################################
+########################################################################################################################
+# COPY VARIOUS SCRIPTS USED BY INSTRUCTORS AND SYS ADMINS FOR COURSE ADMINISTRATION
+
+echo -e "Copy the user scripts"
+
+if [ -z ${SUBMITTY_INSTALL_DIR+x} ]; then
```
Comment (bmcutler) [inline .setup/INSTALL_SUBMITTY_HELPER_BIN.sh]: is the "+x" a piece of magic bash syntax?

File: .setup/INSTALL_SUBMITTY_HELPER_BIN.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (MasterOdin) [reply .setup/INSTALL_SUBMITTY_HELPER_BIN.sh]: the `-z` tests for not empty (null or otherwise). The variable is then read as `${VARIABLE+[word]}` which is that if `${VARIABLE}` is null/empty, replace it with nothing, else replace it with `[word]` (or in this case, `x`).

File: .setup/INSTALL_SUBMITTY_HELPER_BIN.sh (hunk 4 above)
Comment (MasterOdin) [reply .setup/INSTALL_SUBMITTY_HELPER_BIN.sh]: Yeah. Copying my answer from above:

The `-z` tests for not set or empty. The variable is then read as `${VARIABLE+[word]}` which is that if `${VARIABLE}` is null/empty, replace it with nothing, else replace it with `[word]` (or in this case, `x`). The check itself is so that if the variable isn't set (because we're running this standalone), then we enter the block and fill in the variables as needed. We do need the `+[word]` because `-z` cannot distinguish between the empty and not set cases.

Basically, I hate the convoluted bash syntax for doing stuff that's super simple in other languages.

Comment (bmcutler) [reply .setup/INSTALL_SUBMITTY_HELPER_BIN.sh]: ok, sorry i missed your answer.  argh to both bash & github.

**Comment by MasterOdin** [comment]
@bmcutler, ugh, missed a bunch of references to bin that should have sbin. Should be good now:

```
root@vagrant:/usr/local/submitty# ls -la sbin
total 144
drwxr-xr-x  3 root root             4096 May 18 22:53 .
drwxr-x--x 16 root course_builders  4096 May 18 23:01 ..
-r-x------  1 root root             6119 May 18 16:40 adduser.py
-r-x------  1 root root             3610 May 18 16:40 anonymize_autograding_logs.py
-r-x------  1 root root             1485 May 18 16:40 anonymize.py
-r-xr-x---  1 root www-data         8175 May 18 16:40 authentication.py
drwxr-x---  2 root hwcron           4096 May 18 16:40 autograder
-r-xr-x---  1 root hwcron           2567 Apr  7 12:58 build_config_upload.py
-r-x------  1 root root             1413 May 18 16:40 build_course.sh
-r-x------  1 root root             7529 May 18 16:40 check_everything.py
-r-x------  1 root root             2879 May 18 16:40 compare_reuploads.py
-r-x------  1 root root             5786 May 18 16:40 count_autograding_logs.py
-r-x------  1 root root            13521 May 18 16:40 create_course.sh
-r-x------  1 root root             6428 May 18 16:40 get_version_details.py
-r-xr-xr-x  1 root root              887 May 18 16:40 killall.py
-r-x------  1 root root             4994 May 18 16:40 replay_experiment_script.sh
-r-x------  1 root root             1729 May 18 16:40 setcsvfields.py
-r-xr-x---  1 root hwcron          26250 May 18 16:40 submitty_autograding_shipper.py
-r-xr-x---  1 root hwcron           7386 May 18 16:40 submitty_autograding_worker.py
-r-x------  1 root root             1780 May 18 16:40 untrusted_canary.py
root@vagrant:/usr/local/submitty# ls -la sbin/autograder/
total 72
drwxr-x--- 2 root hwcron  4096 May 18 16:40 .
drwxr-xr-x 3 root root    4096 May 18 22:53 ..
-r-xr-x--- 1 root hwcron 40222 May 18 16:40 grade_item.py
-r-xr-x--- 1 root hwcron  1353 May 18 16:40 grade_items_logging.py
-r-xr-x--- 1 root hwcron   105 May 18 16:40 __init__.py
-r-xr-x--- 1 root hwcron 11197 May 18 16:40 insert_database_version_data.py
-r-xr-x--- 1 root hwcron  2107 May 18 16:40 write_grade_history.py
root@vagrant:/usr/local/submitty# ls -la bin
total 1084
drwxr-xr-x  2 root   course_builders   4096 May 18 23:01 .
drwxr-x--x 16 root   course_builders   4096 May 18 23:01 ..
-r-xr-x---  1 hwcron course_builders   5490 May 18 16:40 build_homework_function.sh
-r-x------  1 root   root              5231 May 18 16:40 generate_repos.py
-r-xr-x---  1 root   course_builders   4672 May 18 16:40 grading_done.py
-r-xr-x---  1 root   course_builders   2709 May  6 12:03 left_right_parse.py
-r-xr-x---  1 hwcron course_builders   2475 May  6 12:03 make_assignments_txt_file.py
-r-xr-x---  1 root   course_builders   2968 May  6 12:03 read_iclicker_ids.py
-r-xr-x---  1 root   course_builders  10213 May 18 16:40 regrade.py
-r-xr-x---  1 root   course_builders 118776 May 18 23:01 system_call_check.out
-r-sr-x---  1 root   hwcron          930136 May 18 23:01 untrusted_execute
```

**Comment by MasterOdin** [comment]
Moved untrusted_execute and went through a double checked everything else again.
